Column management for Pods Page

jcaianirh opened the pull request:
For: https://issues.redhat.com/browse/CONSOLE-2239

Todo:

- [x] Further test

- [x] Add settings to local storage

- [x] Fix issue with some subnav pod tables

- [x] Unit test

- [ ] e2e test

![col-mgmnt](https://user-images.githubusercontent.com/35978579/85329135-f1568380-b49f-11ea-909e-9c843695f0a7.gif)

@openshift/team-ux-review 

Thanks @bipuladh for help getting over the redux state change issue

spadgett commented:
> [ ]  e2e test

I'd suggest we target Cypress (#4831) for any e2e tests as the CRUD tests are moving there. We might see if there is a way to test this generically for tables that implement column management.

spadgett commented on frontend/public/components/_autocomplete.scss:
```diff
@@ -2,7 +2,6 @@
   background-color: #fff;
   z-index: 5;
   width: 100%;
-  max-width: 270px;
```
Do we know why this style was here previously and what other pages this might affect?

jcaianirh replied:
This was just added by bipul for the column filter component. I verified this doesn't change the other filters adversely

spadgett commented on frontend/public/actions/ui.ts:
```diff
@@ -281,6 +283,13 @@ export const sortList = (
 };
+export const setColumnManagementFilter = (id: string, filter: any) => {
+  const currentFilters =
+    JSON.parse(localStorage.getItem(COLUMN_MANAGEMENT_LOCAL_STORAGE_KEY)) ?? {};
```
Need to handle parse errors.

jcaianirh replied:
@spadgett ok, if we run into a parsing error retrieving the filters from the store, i'm now logging an error message to the console, and aborting the operation.

spadgett commented on frontend/public/actions/ui.ts:
```diff
@@ -18,6 +19,7 @@ import { OverviewSpecialGroup } from '../components/overview/constants';
 import { setClusterID, setCreateProjectMessage, setUser, setConsoleLinks } from './common';
 
 export enum ActionType {
+  SetColumnManagementFilter = 'setColumnManagementFilter',
```
Is this really a "filter"? Maybe `setTableColumns`?

jcaianirh replied:
@spadgett i was thinking its similar to our existing rowFilters, instead we are filtering the view from the column perspective, but I can change to setTableColumns.

spadgett commented on frontend/packages/console-shared/src/constants/common.ts:
```diff
@@ -36,6 +36,7 @@ export const API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY = `${STORAGE_PREFIX}/api-
 export const COMMUNITY_PROVIDERS_WARNING_LOCAL_STORAGE_KEY = `${STORAGE_PREFIX}/community-providers-warning`;
 export const DEV_CATALOG_FILTER_KEY = `${STORAGE_PREFIX}/dev-catalog-filters`;
 export const PINNED_RESOURCES_LOCAL_STORAGE_KEY = `${STORAGE_PREFIX}/pinned-resources`;
+export const COLUMN_MANAGEMENT_LOCAL_STORAGE_KEY = `${STORAGE_PREFIX}/column-management`;
```
Maybe call this "table columns" instead of "column management"? I think that makes it more clear what it's for.

```suggestion
export const TABLE_COLUMNS_LOCAL_STORAGE_KEY = `${STORAGE_PREFIX}/table-columns`;
```

jcaianirh replied:
done

spadgett commented on frontend/public/components/_filter-toolbar.scss:
```diff
@@ -23,6 +23,6 @@
-    flex-grow: 1;
-    max-width: 90vw;
+  flex-grow: 1;
+  max-width: 370px;
```
I think we at least need to change the name of this class if it no longer fills the width. Do we need to restrict it so much? It's a shame we're losing space to type filters.

jcaianirh replied:
@spadgett we don't need to restrict it. Without any css changes, the container of the filter was 90vw which was putting the table column icon to the far right even though the filter wasn't filling up all that horizontal space. (see the pic) 

This simply keeps the container the same size as the component, we are not losing any filter text.

![Screen Shot 2020-07-10 at 3 45 57 PM](https://user-images.githubusercontent.com/35978579/87192448-77f2c980-c2c4-11ea-93d9-987d43fc0203.png)

spadgett commented on frontend/public/components/factory/list-page.jsx:
```diff
@@ -59,17 +59,20 @@ export const TextFilter = (props) => {
       textFilter,
       hideToolbar,
       hideLabelFilter,
+      hideColumnFilter,
+      columnFilters,
```
Again, I don't think these are really filters. Maybe `selectedColumns`.

jcaianirh replied:
updated to selectedColumns

spadgett commented on frontend/public/components/factory/modal.tsx:
```diff
@@ -118,15 +118,27 @@ export const ModalSubmitFooter: React.SFC<ModalSubmitFooterProps> = ({
   return (
     <ModalFooter inProgress={inProgress} errorMessage={errorMessage} message={message}>
       <ActionGroup className="pf-c-form pf-c-form__actions--right pf-c-form__group--no-top-margin">
+        {resetText && (
```
It would be better to check for `reset` and then just have generic `Reset` as the text if `resetText` was not provided.

jcaianirh replied:
done

spadgett commented on frontend/public/components/factory/table.tsx:
```diff
@@ -397,6 +407,7 @@ export type TableProps = {
   reduxID?: string;
   reduxIDs?: string[];
   label?: string;
+  kind?: string; // for column management
```
Do we ever want to add this to tables that aren't resource tables? We might want to use something other than `kind`.

If this is just the `kind` string, it wouldn't be unique. We should make sure we're using API group + kind as the key.

We should decide if we want to include API version here or not. The problem is that the selected columns would be forgotten if we ever update API versions. However we might need to change the table columns anyway if the API changes.

jcaianirh replied:
for now using gvk as a unique id and passing in an additional col management type string for user facing strings

spadgett commented on frontend/public/components/factory/table.tsx:
```diff
@@ -415,13 +426,23 @@ type ComponentProps = {
+const getActiveColumns = (Header, componentProps, activeColumns) => {
+  let columns = Header(componentProps);
+  if (!_.isEmpty(activeColumns)) {
+    columns = columns?.filter((col) => activeColumns?.[col.title] || col.title === '');
```
I would use something other than `title` to identify the column. We probably want an internal ID for each.

We need to consider how this will work if we translate the UI.

jcaianirh replied:
Good idea, added an ID.

spadgett commented on frontend/public/components/factory/table.tsx:
```diff
@@ -455,6 +476,7 @@ export const Table = connect<
       sortList: PropTypes.func,
       onSelect: PropTypes.func,
       scrollElement: PropTypes.oneOf([PropTypes.object, PropTypes.func]),
+      kind: PropTypes.string, // for column management should use gvk
```
Call this something other than `kind` if it's gvk.

It might be better to give this a specific property like `columnManagementID` or something to make it really clear what's it for. Then you don't need the comment.

jcaianirh replied:
replaced kind with columnManagementID throughout. Thanks!

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+const OVERRIDES = ['Name'];
+
+const getDataListRows = (
+  rowData: any[],
```
Can we use better types than `any[]`?

jcaianirh replied:
rowData no longer being used.

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+
+const getDataListRows = (
+  rowData: any[],
+  onChange: (checked: boolean, event: any) => void,
```
Better type than `any`?

jcaianirh replied:
updated all event types to use React.SyntheticEvent

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+  disableUncheckedRows: boolean,
+): React.ReactNode[] =>
+  rowData.map((column) => (
+    <DataListItem aria-labelledby="table-column-management-item1" key={column.title}>
```
I don't see an element with ID `table-column-management-item1`

jcaianirh replied:
fixed to add dynamic labelled by strings based on unique ID

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+  );
+
+  const updateColumns = (checked: boolean, name: string, columnArray: any): any => {
+    const updatedColumns = _.cloneDeep(columnArray);
+    return updatedColumns.map((column) => {
+      if (_.isEmpty(name) || column.title === name) {
+        column.visible = checked;
+        return column;
+      }
+      return column;
+    });
```
```suggestion
    return updatedColumns.map((column) => {
      if (!name || column.title === name) {
        return { ...column, visible: checked };
      }
      return column;
    });
```

jcaianirh replied:
thanks updated

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+}) => {
+  const dispatch = useDispatch();
+  const initialDefaultColumns = !_.isEmpty(columnFilters)
+    ? _.cloneDeep(columnFilters).filter((column) => column.title.length > 0 && !column.additional)
```
No need for `cloneDeep`. `filter` already creates a new array.

```suggestion
    ? columnFilters.filter((column) => column.title.length > 0 && !column.additional)
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter

jcaianirh replied:
done

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+    ? _.cloneDeep(columnFilters).filter((column) => column.title.length > 0 && !column.additional)
+    : {};
+  const initialAdditionalColumns = !_.isEmpty(columnFilters)
+    ? _.cloneDeep(columnFilters).filter((column) => column.additional)
```
```suggestion
    ? columnFilters.filter((column) => column.additional)
```

jcaianirh replied:
done

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+    initialAdditionalColumns,
+  );
+
+  const updateColumns = (checked: boolean, name: string, columnArray: any): any => {
```
Use a better type than `any`.

spadgett replied:
I'd suggest storing checked state internally as a map and only converting to an array when saved. Then you don't need to complicated array manipulation on every change.

jcaianirh replied:
@spadgett not sure how the map will help make things less complex here...also not sure its super complicated...i tried converting, but it seemed to make things more difficult. Maybe i'm missing something here.

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+    });
+  };
+
+  const defaultColumnChange = (checked: boolean, event): void => {
```
Add a type for `event` (here and below).

jcaianirh replied:
all set

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+
+  const areMaxColumnsDisplayed =
+    additionalColumns.filter((column) => column.visible).length +
+      defaultColumns.filter((column) => column.visible).length ===
```
```suggestion
      defaultColumns.filter((column) => column.visible).length >=
```

jcaianirh replied:
done.

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+          <Alert
+            className="co-alert"
+            isInline
+            title={`You can select up to ${MAX_VIEW_COLS - 1} columns`}
```
I assume we're subtracting one here because name is disabled? Will that be confusing because we show name as selected in the dialog?

jcaianirh replied:
removing the subtraction i think that might be more clear.

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+            <span className="col-sm-6">
+              <label className="control-label">{`Default ${kinds[0]} Columns`}</label>
+              <DataList aria-label="default column list" id="defalt-column-management" isCompact>
+                {getDataListRows(defaultColumns, defaultColumnChange, areMaxColumnsDisplayed)}
```
Make this a proper React component instead of a function that returns markup.

jcaianirh replied:
done.

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+export type ColumnManagementModalProps = {
+  cancel?: () => void;
+  close?: () => void;
+  kinds: any;
```
I'm unclear what `kinds` is here.

jcaianirh replied:
changed to columnManagementID

spadgett commented on frontend/public/components/pod.tsx:
```diff
@@ -104,15 +109,40 @@ const tableColumnClasses = [
   classNames('pf-m-hidden', { 'pf-m-visible-on-xl pf-u-w-10-on-2xl': showMetrics }),
   classNames('pf-m-hidden', { 'pf-m-visible-on-xl pf-u-w-10-on-2xl': showMetrics }),
   classNames('pf-m-hidden', 'pf-m-visible-on-2xl pf-u-w-10-on-2xl'),
+  classNames('pf-m-hidden', 'pf-m-visible-on-xl'),
+  classNames('pf-m-hidden', 'pf-m-visible-on-xl'),
+  classNames('pf-m-hidden', 'pf-m-visible-on-xl'),
```
It seems like it should no longer be possible to use classes like `pf-m-visible-on-xl` for table columns. If I only select two columns, those columns should always be visible no matter what my screen width is.

jcaianirh replied:
@spadgett good point...not sure what this css would be for this case...do we do this in other tables that you are aware of?

spadgett replied:
> @spadgett good point...not sure what this css would be for this case...do we do this in other tables that you are aware of?

I think we need to cede control to the list view and let it calculate visibility based on screen width and the index of the visible columns. At each screen breakpoint, we show a certain number of columns. That has the potential to clean up the list views a ton since we can get rid of the CSS in most cases. You'd only need to set the CSS in special cases for when you need to control column width.

spadgett commented on frontend/public/components/pod.tsx:
```diff
@@ -104,15 +109,40 @@ const tableColumnClasses = [
+  children,
+}) => {
+  return (
+    ((!_.isEmpty(columnFilter)
+      ? columnFilter?.find((col) => col.title === columnName && col.visible === true)
+      : true) && <TableData className={className}>{children}</TableData>) ||
```
```suggestion
    const visible = _.isEmpty(columnFilter) || columnFilter?.find((col) => col.title === columnName && col.visible === true);
    return visible ? <TableData className={className}>{children}</TableData> : null;
```

spadgett replied:
We should generalize this so we don't need this logic in every list view.

jcaianirh replied:
ok, removed need for PodTableData and added all necessary logic into TableData to generalize it.

spadgett commented on frontend/public/components/pod.tsx:
```diff
@@ -181,58 +268,92 @@ const getHeader = (showNodes) => {
         sortField: 'metadata.name',
         transforms: [sortable],
         props: { className: tableColumnClasses[0] },
+        visible: true,
```
We shouldn't need to define a `visible` prop here imo. We should calculate it from the column settings. If there are no column settings, you simply check `!additional`.

jcaianirh replied:
@spadgett i think you're right...since this only drives the default column list for the cm dialog, any user modifications will be stored in state/local storage. At that point we will need to know visibility and additional, but not here.

jcaianirh replied:
@spadgett I take that back, i had a problem defining default state in the initial pr. On this round of edits, Visible and additional are two properties that define the default state of the component and feeds into the col management modal for position in the two tables. I'm going to need them both after trying to get away from using visible.

spadgett replied:
Can you expand on the problem? I think we can make it work, and it would be a cleaner API.

spadgett commented on frontend/public/reducers/ui.ts:
```diff
@@ -85,6 +86,8 @@ export default (state: UIState, action: UIAction): UIState => {
     const storedPins = localStorage.getItem(PINNED_RESOURCES_LOCAL_STORAGE_KEY);
     const pinnedResources = storedPins ? JSON.parse(storedPins) : {};
+    const storedColumnFilters =
+      JSON.parse(localStorage.getItem(COLUMN_MANAGEMENT_LOCAL_STORAGE_KEY)) ?? {};
```
Handle parse errors

jcaianirh replied:
done

spadgett commented:
We should look at how we decide what columns are visible are what window widths. It needs to be more dynamic that the static classes we use today.

We should also decide what happens when columns change in the future (when I add or remove a column to an existing list view). The behavior is not clear to me when columns have been customized.

jcaianirh commented:
> We should also decide what happens when columns change in the future (when I add or remove a column to an existing list view). The behavior is not clear to me when columns have been customized.

@spadgett 

- To add a column to a table without column management there is no difference in a list view.

- To enable column management on a workload
Add the columnManagementID to the Table so it can assess visibility of columns in global state

Add columnMangementID, Type, and selectedColumns (either the default state or user defined in global state) to ListView which passes the data to the modal to allow for col mgmnt configuration by user.

In the header, add a column ID, visible, and additional states to provide a default config for showing or hiding columns.

Add the column visibility state, and ID to the TableData for or hiding cell data.

- To add a column to a workload already enabled with col management
Update the TableData and header info with the visibility state above including weather additional column and default visibility.

jcaianirh commented:
@spadgett i addressed most of these edits and rebased. I left open the 3 comments I had questions on and marked the rest resolved.

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,186 @@
+};
+
+type ColumnHeaderRow = {
+  ID: string;
```
Lowercase `id`

```suggestion
  id: string;
```

jcaianirh replied:
Removed the need for this type altogether.

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,186 @@
+export const MAX_VIEW_COLS = 9;
+
+export const NAME_COLUMN_ID = 'NAME';
+const READ_ONLY_COLS = [NAME_COLUMN_ID];
```
Make this a `Set` instead of an array

jcaianirh replied:
done

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,186 @@
+}) => {
+  const dispatch = useDispatch();
+  const initialDefaultColumns = !_.isEmpty(selectedColumns)
+    ? selectedColumns.filter((column) => column.title.length > 0 && !column.additional)
```
```suggestion
    ? selectedColumns.filter((column) => column.title && !column.additional)
```

jcaianirh replied:
@spadgett needed this syntax for kebab columns before your last review, however, since we've added id's, i can make this more clear by checking for presence of column.id now

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,186 @@
+  const dispatch = useDispatch();
+  const initialDefaultColumns = !_.isEmpty(selectedColumns)
+    ? selectedColumns.filter((column) => column.title.length > 0 && !column.additional)
+    : {};
```
You're calling `filter` above. Should this be an empty array?

We need better types here. It's declared as `any` below.

jcaianirh replied:
@spadgett yes its an array of objects, so this should be []. thanks...working on types.

jcaianirh replied:
@spadgett added ManagedColumn type to better support typing where needed.

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,186 @@
+    : {};
+  const initialAdditionalColumns = !_.isEmpty(selectedColumns)
+    ? selectedColumns.filter((column) => column.additional)
+    : {};
```
Same question here. Is this really an array?

jcaianirh replied:
@spadgett its an array thanks

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,186 @@
+  const initialAdditionalColumns = !_.isEmpty(selectedColumns)
+    ? selectedColumns.filter((column) => column.additional)
+    : {};
+  const [defaultColumns, setDefaultColumns] = React.useState<any[]>(initialDefaultColumns);
```
Can you specify a better type than `any`?

jcaianirh replied:
@spadgett  yes, fixed up types and added ManagedColumn type. (as an aside these are no longer using state, but did fix other types)

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,186 @@
+    initialAdditionalColumns,
+  );
+
+  const updateColumns = (checked: boolean, id: string, columnArray: any): any => {
```
Improve types

spadgett replied:
I think the logic would be simpler if the selected columns were stored in state as a set, and then you update the values in the array on save.

jcaianirh replied:
@spadgett i was looking at Map last time...do you have a pref between set and map? I can look more into this.

jcaianirh replied:
@spadgett took your advice, moved all column logic to use a set, removed this function entirely, and simplified the logic on the page by eliminating these array operations.

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,186 @@
+ColumnManagementModal.displayName = 'ColumnManagementModal';
+
+type DataListRowProps = {
+  column: any;
```
Improve types

jcaianirh replied:
@spadgett improved types.

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,186 @@
+  cancel?: () => void;
+  close?: () => void;
+  columnManagementID: string;
+  selectedColumns: any;
```
Improve types

jcaianirh replied:
done

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,186 @@
+        <div className="co-m-form-row">
+          <p>Checked categories will be displayed in the table.</p>
+        </div>
+        <div className="row co-m-form-row">
```
Is `row` needed here?

jcaianirh replied:
changes spacing a bit but not needed here. more in line with the design without it...removed. thanks

spadgett commented on frontend/public/components/pod.tsx:
```diff
@@ -104,140 +109,209 @@ const tableColumnClasses = [
       },
       {
         title: 'Namespace',
+        ID: 'NAMESPACE',
```
Prefer lowercase to all caps

```suggestion
        id: 'namespace',
```

jcaianirh replied:
done

spadgett commented on frontend/public/components/pod.tsx:
```diff
@@ -104,140 +109,209 @@ const tableColumnClasses = [
+        >
+          <ResourceLink kind="Namespace" name={namespace} />
+        </TableData>
+        <TableData className={tableColumnClasses[2]} columns={columns} columnID="STATUS">
```
It would be good to find a way to avoid repeating the IDs for every column. I'm worried there will be typos and it will be hard to maintain.

jcaianirh replied:
created an object to store repetitive column data (col classes, id's, and titles). reference that for each reference of these properties.

spadgett commented on frontend/public/components/pod.tsx:
```diff
@@ -644,6 +727,7 @@ type PodTableRowProps = {
 
 type PodTableRowPropsFromState = {
   metrics: UIActions.PodMetrics;
+  selectedColumns: any;
```
Improve types

jcaianirh replied:
done

jcaianirh commented:
@spadgett updated code based on resolved comments above. working on unresolved comments.

jcaianirh commented:
@spadgett updated code for type review comments

spadgett commented on frontend/__tests__/components/modals/column-management-modal.spec.tsx:
```diff
@@ -11,71 +11,71 @@ import {
 import { PodModel } from '@console/internal/models';
 
-const columnManagementID = referenceForModel(PodModel);
+const columnManagementId = referenceForModel(PodModel);
```
The rule is `id` all lowercase if it's the first word, but `myID` with uppercase `ID` if a later word of a camel case variable name.

```suggestion
const columnManagementID = referenceForModel(PodModel);
```

jcaianirh replied:
ok, fixed.

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -47,55 +54,53 @@ const DataListRow: React.FC<DataListRowProps> = ({ column, onChange, disableUnch
-      return column;
-    });
-  };
+  const [checkedColumns, setCheckedColumns] = React.useState<Set<string>>(new Set(initialSelected));
```
nit: TypeScript can infer the `useState` type as long as it's passed a value.

```suggestion
  const [checkedColumns, setCheckedColumns] = React.useState(new Set(initialSelected));
```

jcaianirh replied:
removed type here

spadgett commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -47,55 +54,53 @@ const DataListRow: React.FC<DataListRowProps> = ({ column, onChange, disableUnch
+      setTableColumns(
+        columnManagementId,
+        selectedColumns.map((column) => {
+          return { ...column, visible: checkedColumns.has(column.id) };
```
How much data about the column are we storing? I'm worried about upgrades here where things like the column headings change. Really we should only be storing a `Set` of selected columns IDs. Everything else about the column should be passed to the table or column managed modal by code.

I think this relates to the need for the `visible` property. The column definition itself should never change, and the visibility should be tracked separately in `localStorage` / Redux.

jcaianirh replied:
@spadgett removed visible prop, only tracked the set of col id's in localstorage/redux. passed only the col props needed to col management modal.

jcaianirh commented:
@spadgett @alimobrem @serenamarie125 @beanh66 @gdoyle1 

Added logic for showing namespace when all namespaces selected using the cm code. Added an example for Deployment and Deployment Config tables to show the logic which is to add the id='namespace' property for the table data and column data for the workloads.

![namespace](https://user-images.githubusercontent.com/35978579/88594677-a1835300-d02f-11ea-85a3-86e1dc26733f.gif)

spadgett commented on frontend/public/components/pod.tsx:
```diff
@@ -582,17 +748,23 @@ export const PodsPage = connect<{}, PodPagePropsFromDispatch, PodPageProps>(
       return () => clearInterval(id);
     }
-  }, [namespace]);
+  }, [namespace, setPodMetrics]);
```
I'm not sure this new dependency is right. Will this cause an infinite loop?

jcaianirh replied:
@spadgett i didn't intentionally change this code...maybe a merge issue...removing as its unintentional.

spadgett commented on frontend/public/components/pod.tsx:
```diff
@@ -121,45 +287,104 @@ const PodTableRow = connect<PodTableRowPropsFromState, null, PodTableRowProps>(p
-        <TableData className={classNames(tableColumnClasses[1], 'co-break-word')}>
+        <TableData
+          className={classNames(podColumnInfo.namespace.classes, 'co-break-word')}
+          columns={columns}
```
It's too bad we have to pass `columns` through to every single table cell. Is there a way to avoid that? Can `TableData` connect directly to Redux? Or maybe through `useContext`?

spadgett commented on frontend/public/actions/ui.ts:
```diff
@@ -282,6 +284,20 @@ export const sortList = (
+    // Error parsing the data, do not store the current filters
+    /* eslint-disable-next-line no-console */
+    console.error('Error parsing column filters from local storage', e);
+    return;
```
We might want to just default to `{}` and reset the columns on parse errors. Otherwise users will never be able to recover?

jcaianirh replied:
@spadgett should i return an new ImmutableMap here? then it will be empty, and columns will view as if no col management was provided.

jcaianirh replied:
assigned currentColumns = {};which will allow recovery and reset col management settings.

jcaianirh commented on frontend/public/components/modals/column-management-modal.tsx:
```diff
@@ -0,0 +1,186 @@
+        <div className="row co-m-form-row">
+          <div className="col-sm-12">
+            <span className="col-sm-6">
+              <label className="control-label">{`Default ${columnManagementType} Columns`}</label>
```
ok, added a single prop here called columnLayout instead of 3. removed string template

jcaianirh commented on frontend/public/components/factory/table.tsx:
```diff
@@ -264,13 +267,27 @@ export type TableRowProps = {
 export type TableDataProps = {
-  id?: string;
   className?: string;
+  columnId?: string;
```
udpated to columnID

jcaianirh commented on frontend/public/components/factory/table.tsx:
```diff
@@ -264,13 +267,27 @@ export type TableRowProps = {
+  ...props
+}) => {
+  const showNamespace =
+    columnId === 'namespace' ? UIActions.getActiveNamespace() === ALL_NAMESPACES_KEY : true;
```
done.

jcaianirh commented on frontend/public/components/factory/table.tsx:
```diff
@@ -424,13 +442,36 @@ type ComponentProps = {
+  }
+
+  const showNamespace = UIActions.getActiveNamespace() === ALL_NAMESPACES_KEY;
+  columns = columns.filter((column) => (column.id === 'namespace' ? showNamespace : true));
```
done thanks for this

jcaianirh commented on frontend/public/components/factory/table.tsx:
```diff
@@ -424,13 +442,36 @@ type ComponentProps = {
+  let columns = Header(componentProps);
+  if (_.isEmpty(activeColumns) && columns.find((col) => col.additional)) {
+    activeColumns = new Set(
+      columns.reduce((acc, curr) => {
+        if (curr.id && !curr.additional) {
+          acc.push(curr.id);
+        }
+        return acc;
+      }, []),
```
@spadgett i knew i went the long way round to get there...

jcaianirh replied:
@spadgett i actually only want the id, not the entire object. the reduce allows me to push the id's to an accumulator.

jcaianirh replied:
changed it to this:`columns.filter((col) => col.id && !col.additional).map((filteredCol) => filteredCol.id)`

jcaianirh commented on frontend/public/components/factory/table.tsx:
```diff
@@ -424,13 +442,36 @@ type ComponentProps = {
   kindObj?: K8sResourceKindReference;
 };
 
+const getActiveColumns = (Header, componentProps, activeColumns) => {
```
added types here.

jcaianirh commented on frontend/public/components/pod.tsx:
```diff
@@ -94,25 +99,186 @@ export const menuActions = [
-  classNames('pf-m-hidden', 'pf-m-visible-on-2xl pf-u-w-10-on-2xl'),
-  Kebab.columnClass,
-];
+const podColumnInfo = {
```
good idea thanks

gdoyle1 commented:
@spadgett @jcaianirh Based on the new logic for hiding namespaces, is the gif that was shown all the way up top accurate? I'm specifically wondering when in all projects, the namespace checkbox in the modal

jcaianirh commented on frontend/public/components/pod.tsx:
```diff
@@ -564,7 +725,12 @@ export const PodsPage = connect<{}, PodPagePropsFromDispatch, PodPageProps>(
+    useSelector<RootState, string>(({ UI }) => UI.getIn(['columnManagement', columnManagementID])),
+  );
+  if (_.isEmpty(selectedColumns)) {
+    selectedColumns = getSelectedColumns(props?.customData?.showNodes);
```
Yes, this is going to the col management model, so we can use other logic to generate the modal and its datalists. Updated the code to not need the fallback else in workload page

spadgett commented on frontend/public/actions/ui.ts:
```diff
@@ -282,6 +284,20 @@ export const sortList = (
+    // Error parsing the data, do not store the current filters
+    /* eslint-disable-next-line no-console */
+    console.error('Error parsing column filters from local storage', e);
+    currentColumns = {};
```
I don't think this is necessary since it was initialized to `{}` above.

jcaianirh replied:
removed this

spadgett commented on frontend/public/components/factory/list-page.jsx:
```diff
@@ -70,6 +70,10 @@ export class ListPageWrapper_ extends React.PureComponent {
       hideLabelFilter,
+      selectedColumns,
+      columnManagementID,
+      columnManagementType,
```
Are these still needed?

jcaianirh replied:
removed these. thanks.

spadgett commented on frontend/public/components/factory/table.tsx:
```diff
@@ -264,13 +267,27 @@ export type TableRowProps = {
+}) => {
+  const showNamespace =
+    columnID !== 'namespace' || UIActions.getActiveNamespace() === ALL_NAMESPACES_KEY;
+  return (
+    ((!_.isEmpty(columns) ? columns.has(columnID) : true) && showNamespace && (
+      <td {...props} className={className} role="gridcell" />
+    )) ||
+    null
+  );
```
```suggestion
  return (
    (_.isEmpty(columns) || columns.has(columnID) && showNamespace
      ? <td {...props} className={className} role="gridcell" />
      : null
  );
```

jcaianirh replied:
This doesn't properly filter out the namespaces for non col management tables...keeping orig return

spadgett replied:
> This doesn't properly filter out the namespaces for non col management tables...keeping orig return

I think I just missed a closing paren. Does this work?

```suggestion
  return (
    (_.isEmpty(columns) || columns.has(columnID)) && showNamespace
      ? <td {...props} className={className} role="gridcell" />
      : null
  );
```

spadgett commented on frontend/public/components/factory/table.tsx:
```diff
@@ -424,13 +442,39 @@ type ComponentProps = {
+  activeColumns: Set<string>,
+) => {
+  let columns = Header(componentProps);
+  if (_.isEmpty(activeColumns) && columns.find((col) => col.additional)) {
```
Is the `find` call here needed? It seems like we should leave that out.

jcaianirh replied:
this is correct, its not needed. removed.

spadgett commented on frontend/public/components/factory/table.tsx:
```diff
@@ -424,13 +442,39 @@ type ComponentProps = {
+  }
+
+  const showNamespace = UIActions.getActiveNamespace() === ALL_NAMESPACES_KEY;
+  columns = columns.filter((column) => column.id !== 'namespace' || showNamespace);
```
nit: you can skip iterating over the columns if `showNamespace` is true

```suggestion
  if (!showNamespace) {
    columns = columns.filter((column) => column.id !== 'namespace');
  }
```

jcaianirh replied:
good point less churn.

jcaianirh commented:
@spadgett made final edits. made a note on one that i didn't change as it didn't provide the desired result.

spadgett commented on frontend/public/actions/ui.ts:
```diff
@@ -282,6 +284,19 @@ export const sortList = (
+  try {
+    currentColumns = JSON.parse(localStorage.getItem(COLUMN_MANAGEMENT_LOCAL_STORAGE_KEY)) || {};
+  } catch (e) {
+    // Error parsing the data, do not store the current filters
```
Comment is no longer accurate

jcaianirh replied:
@spadgett yes this is working correctly...thanks for improving this.

jcaianirh replied:
updated comment

spadgett commented:
/kind feature

jcaianirh commented:
@spadgett made edits. thanks.

jcaianirh commented:
> Thanks @jcaianirh. Please open a bug to fix column visibility at different screen widths. We can see if there is an opportunity to simplify the API when working on that.
> 
> /lgtm

https://bugzilla.redhat.com/show_bug.cgi?id=1862564

jcaianirh commented:
/test e2e-gcp-console

spadgett commented:
/retest
